### PR TITLE
webgpu: change GPUDevice to be an event handler

### PR DESCRIPTION
**Pull request by edevil**
Depends on https://github.com/cloudflare/workerd/pull/994.

Transform GPUDevice into an event handler.

### CODE REVIEW

File: src/workerd/api/gpu/gpu-errors.h [hunk 1]
```diff
@@ -13,19 +13,18 @@ namespace workerd::api::gpu {
 class GPUError : public jsg::Object {
 public:
   explicit GPUError(kj::String m) : message_(kj::mv(m)){};
-  JSG_RESOURCE_TYPE(GPUError) {
-    JSG_READONLY_PROTOTYPE_PROPERTY(message, getMessage);
-  }
+  JSG_RESOURCE_TYPE(GPUError) { JSG_READONLY_PROTOTYPE_PROPERTY(message, getMessage); }
```
Comment (jasnell) [inline src/workerd/api/gpu/gpu-errors.h:16]:
```suggestion
  JSG_RESOURCE_TYPE(GPUError) {
    JSG_READONLY_PROTOTYPE_PROPERTY(message, getMessage);
  }
```

File: src/workerd/api/gpu/gpu-errors.h [hunk 2]
```diff
@@ -13,19 +13,18 @@ namespace workerd::api::gpu {
...
   kj::String message_;
-  kj::StringPtr getMessage() { return message_; }
 };
 
-class GPUOOMError : public GPUError {
+class GPUOutOfMemoryError : public GPUError {
 public:
   using GPUError::GPUError;
-  JSG_RESOURCE_TYPE(GPUOOMError) { JSG_INHERIT(GPUError); }
+  JSG_RESOURCE_TYPE(GPUOutOfMemoryError) { JSG_INHERIT(GPUError); }
```
Comment (jasnell) [inline src/workerd/api/gpu/gpu-errors.h:27]:
```suggestion
  JSG_RESOURCE_TYPE(GPUOutOfMemoryError) {
    JSG_INHERIT(GPUError);
  }
```

File: src/workerd/api/gpu/gpu-errors.h [hunk 3]
```diff
@@ -34,6 +33,12 @@ class GPUValidationError : public GPUError {
   JSG_RESOURCE_TYPE(GPUValidationError) { JSG_INHERIT(GPUError); }
 };
 
+class GPUInternalError : public GPUError {
+public:
+  using GPUError::GPUError;
+  JSG_RESOURCE_TYPE(GPUInternalError) { JSG_INHERIT(GPUError); }
```
Comment (jasnell) [inline src/workerd/api/gpu/gpu-errors.h:39]:
```suggestion
  JSG_RESOURCE_TYPE(GPUInternalError) {
    JSG_INHERIT(GPUError);
  }
```

File: src/workerd/api/gpu/gpu-errors.h [hunk 4]
```diff
@@ -34,6 +33,12 @@ class GPUValidationError : public GPUError {
   JSG_RESOURCE_TYPE(GPUValidationError) { JSG_INHERIT(GPUError); }
```
Comment (jasnell) [inline src/workerd/api/gpu/gpu-errors.h:33]:
```suggestion
  JSG_RESOURCE_TYPE(GPUValidationError) {
    JSG_INHERIT(GPUError);
  }
```

File: src/workerd/api/gpu/gpu-device.c++ [hunk 5]
```diff
@@ -112,18 +110,16 @@ wgpu::MipmapFilterMode parseMipmapFilterMode(kj::StringPtr mode) {
...
-  desc.addressModeW = parseAddressMode(
-      descriptor.addressModeW.orDefault([] { return "clamp-to-edge"_kj; }));
-  desc.magFilter = parseFilterMode(
-      descriptor.magFilter.orDefault([] { return "nearest"_kj; }));
-  desc.minFilter = parseFilterMode(
-      descriptor.minFilter.orDefault([] { return "nearest"_kj; }));
-  desc.mipmapFilter = parseMipmapFilterMode(
-      descriptor.mipmapFilter.orDefault([] { return "nearest"_kj; }));
+  desc.addressModeU =
+      parseAddressMode(descriptor.addressModeU.orDefault([] { return "clamp-to-edge"_kj; }));
```
Comment (jasnell) [inline src/workerd/api/gpu/gpu-device.c++:114]: If these are going to just be literal kj::StringPtrs, then using the `descriptor.addressModeU.orDefault("clamp-to-edge"_kj")` form is just fine. You want to use the callback variant only when there's an allocation cost to the default (e.g. like heap allocating a string with `kj::str(...)`)

**Review by jasnell — APPROVED** [review]
LGTM with a handful of minor nits.

File: src/workerd/api/gpu/gpu-device.c++ (hunk 5 above)
Comment (edevil) [reply src/workerd/api/gpu/gpu-device.c++:114]: This is the compiler trolling me.

descriptor.addressModeU.orDefault("clamp-to-edge"_kj)
```
common.h(1460, 6): Candidate function not viable: no known conversion from 'kj::StringPtr' to 'kj::String &' for 1st argument
common.h(1467, 12): Candidate function not viable: no known conversion from 'kj::StringPtr' to 'const kj::String' for 1st argument
```
descriptor.addressModeU.orDefault([] { return kj::str("clamp-to-edge"); })
```
common.h(1460, 6): Candidate function not viable: no known conversion from '(lambda at /Users/acruz/work/vc/workerd/src/workerd/api/gpu/gpu-device.c++:114:58)' to 'kj::String &' for 1st argument
common.h(1467, 12): Candidate function not viable: no known conversion from '(lambda at /Users/acruz/work/vc/workerd/src/workerd/api/gpu/gpu-device.c++:114:58)' to 'const kj::String' for 1st argument
```

Comment (jasnell) [reply src/workerd/api/gpu/gpu-device.c++:114]: ah right, ok.

**Pull request merged**